cmd/boulder-wfe2/main.go

```diff
@@ -214,13 +237,64 @@ func loadCertificateChains(chainConfig map[string][]string, requireAtLeastOneCha
 
 		// Save the full PEM chain contents, if any
 		if buffer.Len() > 0 {
-			results[aiaIssuerURL] = buffer.Bytes()
+			results[id] = buffer.Bytes()
 		}
 	}
 
 	return results, issuerCerts, nil
 }
 
+// loadChain takes a list of filenames containing pem-formatted certificates,
+// and returns a chain representing all of those certificates in order. It
+// ensures that the resulting chain is valid. The final file is expected to be
+// a root certificate, which the chain will be verified against, but which will
+// not be included in the resulting chain.
+func loadChain(certFiles []string) (*issuance.Certificate, []byte, error) {
+	if len(certFiles) < 2 {
+		return nil, nil, errors.New(
+			"each chain must have at least two certificates: an intermediate and a root")
+	}
+
+	firstCert, err := core.LoadCert(certFiles[0])
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load certificate: %w", err)
+	}
+
+	var buf bytes.Buffer
+	var nextCert *x509.Certificate
+
+	// Iterate over all certs except for the last, checking that their signature
+	// comes from the next cert in the list, and appending their pem to the buf.
+	for i := 0; i < len(certFiles)-1; i++ {
+		var cert *x509.Certificate
+		if nextCert != nil {
+			cert = nextCert
+		} else {
+			cert = firstCert
+		}
+
+		nextCert, err = core.LoadCert(certFiles[i+1])
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to load certificate: %w", err)
+		}
+
+		err = cert.CheckSignatureFrom(nextCert)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to verify chain: %w", err)
+		}
+
+		buf.Write([]byte("\n"))
+		buf.Write(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
+	}
+
+	err = nextCert.CheckSignatureFrom(nextCert)
+	if err != nil {
+		return nil, nil, fmt.Errorf("final cert in chain must be self-signed: %w", err)
```

Comment:
It's unusual to require a root in a certificate chain (in the more typical case of configuring TLS servers), so I think we should provide a really nice error message to explain things here. "... must be a self-signed root (which is used only for validation and not sent to the user)"